### PR TITLE
chore(aws_kinesis sinks): refactor `aws_kinesis_firehose` and `aws_kinesis_streams` sinks to extract redundant code

### PR DESCRIPTION
**Pull request by neuronull**
### Changes:

- creates a common `sinks/aws_kinesis` module which contains the common logic to aws_kinesis sinks including the config, service, request_builder, sink.
- introduces the `Record` and `SendRecord` traits which allow for the actual aws kinesis sink implementations to use the common code.

### NOTES:

At first glance if you just look at lines changed, you may be thinking "why the effort for not much reduction in overall code".

Well, looking closer, there is less functional code overall. The shared logic is not repeated in two places anymore, whereas before, 85% of the two directories `aws_kinesis_streams` and `aws_kinesis_firehose` was identical. Now the `streams` and `firehose` directories contain just the unique bits.

TODO

- [ ] Delete the branch  `neuronull/request_metadata_for_driver` after this PR is merged in

### CODE REVIEW

File: src/sinks/aws_kinesis/request_builder.rs [hunk 1]
```diff
@@ -90,7 +95,7 @@ impl RequestBuilder<KinesisProcessedEvent> for KinesisRequestBuilder {
 
     fn split_input(&self, mut event: KinesisProcessedEvent) -> (Self::Metadata, Self::Events) {
         let metadata = KinesisMetadata {
-            finalizers: event.take_finalizers(),
+            finalizers: event.event.take_finalizers(),
```
Comment (spencergilbert) [inline src/sinks/aws_kinesis/request_builder.rs:98]: This is a little gross with the repeated name but c'est la vie

**Review by spencergilbert — APPROVED** [review]
Looks like a reasonable change, just one lil complaint. Always happy to see reduced duplication.

File: src/sinks/aws_kinesis/request_builder.rs (hunk 1 above)
Comment (neuronull) [reply src/sinks/aws_kinesis/request_builder.rs:98]: Ooo yeah that is a bit gross. I'll take care of that :+1:

**Review by spencergilbert — APPROVED** [review]
Shiny!

**Review by spencergilbert — APPROVED** [review]

**Pull request merged**